Update Follow Me documentation reflecting recent refactor in Follow Target (#1810)
PX4/PX4-user_guide: en/flight_modes/follow_me.md

---
author: Jimmy Johnson

> author: Junwoo Hwang

Reviewer: We don't use these keys and the duplicate breaks the build. Propose all this front matter is removed and this is added to acknowledgements heading at the bottom - something like "Thanks to Jimmy Johnson and @junwoo091400". Even that is not strictly necessary.

---
author: Jimmy Johnson and Junwoo Hwang
---

# Follow-Me Mode

[<img src="../../assets/site/position_fixed.svg" title="Position fix required (e.g. GPS)" width="30px" />](../getting_started/flight_modes.md#key_position_fixed)

*Follow Me* mode allows a multicopter to autonomously hold position and altitude relative to another system that is broadcasting its position (and optionally velocity) using the [FOLLOW_TARGET](https://mavlink.io/en/messages/common.html#FOLLOW_TARGET) MAVLink message.

![Follow-me Concept](../../assets/flight_modes/followme_concept.png)

The vehicle will automatically yaw to face and follow the target from a specified [relative angle](#FLW_TGT_FA), [distance](#FLW_TGT_DST) and [height](#FLW_TGT_HT) and altitude, depending on the [altitude control mode](#FLW_TGT_ALT_M).

By default it will follow from directly behind the target at a distance of 8 meters, and a height of 8 meters above the home (arming) position.

> Users can adjust the follow angle, height and distance using a RC controller as shown above:

Reviewer: ```suggestion
Users can adjust the follow angle, height and distance using an RC controller as shown above:

```

- _Follow Height_ is controlled with the `up-down` input ("Throttle").
  Center the stick to keep follow the target at a constant hight. Raise or lower the stick to adjust height.
- _Follow Distance_ is controlled with the `forward-back` input ("Pitch").
  Pushing the stick forward increases the follow distance, pulling it back decreases the distance.
- _Follow Angle_ is controlled with the `left-right` input ("Roll").
  The movement is from the user's perspective, so if you face the drone and move the stick left, it will move to your left.
  From above if you move the stick left the drone will move counter-clockwise.

  Follow Angle is defined as increasing in clockwise direction relative to the target's heading (which is 0 degrees)

  ![Follow-me Angle Diagram](../../assets/flight_modes/followme_angle.png)

:::note

> If you want to reset the RC-adjusted behavior, you need to either switch off from Follow-Me mode and activate again, or change the according [parameter](#parameters) to a different value.

Reviewer: What does this mean? I.e. what behaviour would I reset and how? I looked at those parameters and they don't have anything for changing what RC input does what?

:::

Demo video:

> @[youtube](https://www.youtube.com/watch?v=myPFVJJkm04)

Reviewer: Video is cool. Would have been nice (if you're out again) to demo it following at some other angle relative to you.

> ## How to Use Follow Me

Reviewer: I merged the topics on safety, QGC and MAVSDK under here, which seemed logical. I did consider calling this "Follow Me Targets". But if we do that then safety should be moved out of this. Thoughts?

Author: I think having that 'How to use' section right after the basic concept introduction is good for readability, and like now having QGC / MAVSDK explanation under this section is good for me 👍

By default it will follow from directly behind the target at a distance of 8 meters, and a height of 8 meters above the home (arming) position.

Users can adjust the follow angle, height and distance using an RC controller as shown above:

- _Follow Height_ is controlled with the `up-down` input ("Throttle").
  Center the stick to keep follow the target at a constant hight. Raise or lower the stick to adjust height.
- _Follow Distance_ is controlled with the `forward-back` input ("Pitch").
  Pushing the stick forward increases the follow distance, pulling it back decreases the distance.
- _Follow Angle_ is controlled with the `left-right` input ("Roll").
  The movement is from the user's perspective, so if you face the drone and move the stick left, it will move to your left.
  From above if you move the stick left the drone will move counter-clockwise.

  Follow Angle is defined as increasing in clockwise direction relative to the target's heading (which is 0 degrees)

  ![Follow-me Angle Diagram](../../assets/flight_modes/followme_angle.png)

> :::note
> If you want to reset the RC-adjustment on angle, height and distance, you need to either exit Follow-Me mode and activate it again

Reviewer: I think this is what you mean:
```suggestion
:::note
Angle, height, and distance values set using the RC controller are discarded when you exit follow-me mode.
If you exit Follow-Me mode and activate it again the values will be reset to their defaults.

```

> *Follow Me* mode allows a multicopter to autonomously follow and track another system that is broadcasting its position using the [FOLLOW_TARGET](https://mavlink.io/en/messages/common.html#FOLLOW_TARGET) MAVLink message.

Reviewer: Is it still just multicopter that supports this mode?

Author: Yep, only multicopter

> The vehicle will automatically yaw to face and follow the target from a specified [relative orientation](#NAV_FT_FS), [distance](#NAV_FT_DST) and [height](#NAV_MIN_FT_HT) (reference point : home, ground (with distance sensor), target's GPS altitude, depending on the [altitude control mode](#NAV_FT_ALT_M)).

Reviewer: I'm not sure a reader will get the point about "reference point ; ..."

Author: Removed!

> *Follow Me* mode allows a multicopter to autonomously follow and track another system that is broadcasting its position using the [FOLLOW_TARGET](https://mavlink.io/en/messages/common.html#FOLLOW_TARGET) MAVLink message.

Reviewer: Are you following something if you are leading it? I suspect not, so I'm suggesting we highlight that it is holding position

```suggestion
*Follow Me* mode allows a multicopter to autonomously hold position and altitude relative to another (possibly moving) system that is broadcasting its position using the [FOLLOW_TARGET](https://mavlink.io/en/messages/common.html#FOLLOW_TARGET) MAVLink message.

```

![Follow-me Concept](../../assets/flight_modes/followme_concept.png)

> The vehicle will automatically yaw to face and follow the target from a specified [relative angle](#NAV_FT_FS), [distance](#NAV_FT_DST) and [height](#NAV_FT_HT) and altitude, depending on the [altitude control mode](#NAV_FT_ALT_M)).

Reviewer: These anchors appear to have been removed.

Author: I removed it since it's not relevant now, as technically it's possible to do Terrain tracking as well as GPS Altitude tracking. They don't work so well, but I thought that can be noted further down in the doc!

> By default it will follow from directly behind the target at a distance of 8 meters, and a height of 8 meters above the home(arming) position.

Reviewer: ```suggestion
By default it will follow from directly behind the target at a distance of 8 meters, and a height of 8 meters above the home (arming) position.

```

> User has an option to control the follow angle, height and distance by :
> - Follow Angle : Roll input. Left and right command will be as if you were controlling the drone's roll while facing it, i.e. Left command moves drone around counter clockwise
> - Follow Height : Throttle input. Center the stick to keep follow height constant, increase or decrease to adjust height
> - Follow Distance : Pitch input. Pushing pitch stick up increases the follow distance, pulling it down decreases the distance.

Reviewer: This only works on MC, so we should probably use MC position mode terminology or a diagram - such as http://docs.px4.io/master/en/getting_started/flight_modes.html#position-mode-mc
i.e. to me the roll input is the left-right input :-)

```suggestion
User has an option to control the follow angle, height and distance by :
- _Follow Angle_ is controlled using the `left-right` input ("Roll"). The movement is from the user's perspective, so if you face the drone and move the stick left, it will move to your left. From above if you move the stick left the drone will move counter-clockwise.
- _Follow Height_ is controlled using the `up-down` input ("Throttle"). Center the stick to keep follow height constant, increase or decrease to adjust height
- _Follow Distance_ is controlled using the `forward-back` input ("Pitch"). Pushing the stick forward increases the follow distance, pulling it back decreases the distance.

```

Author: This explanation is much clearer! I struggled on how to explain the follow angle part!
"from user's perspective, if you face the drone ..." part is a great addition :+1:

> Follow Angle is defined as increasing in clockwise direction, and it get's applied relative to where the target is headed:

Reviewer: ```suggestion
Follow Angle is defined as increasing in clockwise direction relative to the target's heading (which is 0 degrees).
```

Author: Applied!

- Be ready to switch back to Position mode, etc, if something goes wrong when using follow me mode for the first time.
  If the target system (e.g. Android Phone)'s GPS is not accurate, this can affect the behavior.

## How to use Follow Me

*Follow Me* mode is supported by *QGroundControl* on ground station hardware that has a GPS module.
The recommended configuration is a USB OTG-capable Android device with two telemetry radios.

To setup *Follow Me* mode:
- Connect a telemetry radio to your ground station device and another to the vehicle (this allows positioning information to be relayed between the two radios).
- Disable sleep-mode on your Android device:
  - This setting can usually be found under: **Settings \> Display**.
  - It is important that you set your Android device to not go to sleep as this could cause the GPS signal to cease being emitted at regular intervals.
- Takeoff to a height of at least 2-3 metres (recommended even though auto-takeoff is supported).
  - Set the vehicle on the ground, press the safety switch and step back at least 10 meters.
  - Arm the vehicle and takeoff.
- Switch into follow me mode.

>   - The copter will first ascend to the specified [minimum height](#NAV_MIN_FT_HT) while facing the target, then move horizontally, to avoid potential collisions

Reviewer: ```suggestion
  - The copter will first ascend to the specified [minimum height](#NAV_MIN_FT_HT) while facing the target, to avoid potential collisions, before moving horizontally, 

```

At this point you can start moving, and the drone will be following you.

The mode has been tested on the following Android devices:
- Nexus 5
- Nexus 7 Tablet

> - Galaxy S10

Reviewer: Suggest you remove the Nexus 5  and put the Galaxy S10 on the top. IF we can test on another recent device we could remove the Nexus 7 too. Upshot, much better to list devices people can get.

## Configuration

:::warning

> Setting the **Altitude mode ([NAV_FT_ALT_M](../advanced_config/parameter_reference.md#NAV_FT_ALT_M)**) to '3D Tracking' can be dangerous and unstable, as Android phones for example are known for having huge altitude errors in their GPS coordinates, around 10 to 50 meters. Therefore, 3D Tracking is only advised if you have an accurate GPS source for the Ground Station!

Reviewer: Perhaps reorder to instruct first:
```suggestion
Do not set the **Altitude mode ([NAV_FT_ALT_M](#NAV_FT_ALT_M)**) to `3D Tracking`!
This can be dangerous and unstable because many Android phones (and other devices) are known to have significant altitude errors in their GPS coordinates (around 10 to 50 meters!). 3D Tracking is only advised if you have an accurate GPS source for the Ground Station!
```

    By *default* this is 8 metres/26 feet above the home (arming) position.
- It is *safer* to manually fly to a safe height before engaging follow-me mode than to engage follow-me mode when landed (even though the mode implements auto take off).
- Give your vehicle sufficient room to stop, especially when it is moving fast.
  Many Android devices do not update their position very frequently, and autopilot estimations of the speed and direction can be inaccurate.
- Be ready to take manual RC control if something goes wrong when using follow me mode for the first time.
  The accuracy of positioning is dependent on the quality of the GPS used by target system.
  If the GPS is not accurate, this will be reflected in follow me.
- Be ready to switch back to Position mode, etc, if something goes wrong when using follow me mode for the first time.
  If the target system (e.g. Android Phone)'s GPS is not accurate, this can affect the behavior.

> ## Using Follow Me with QGroundControl

Reviewer: Unless you are also covering MAVSDK in the instructions, it is probably better to note that this is using QGroundContol.

Note, I prefer the change you are suggesting, just needs to have some comments about what might be different when using it with MAVSDK.

- Disable sleep-mode on your Android device:
  - This setting can usually be found under: **Settings \> Display**.
  - It is important that you set your Android device to not go to sleep as this could cause the GPS signal to cease being emitted at regular intervals.
- Takeoff to a height of at least 2-3 metres (recommended even though auto-takeoff is supported).
  - Set the vehicle on the ground, press the safety switch and step back at least 10 meters.
  - Arm the vehicle and takeoff.
- Switch into follow me mode.
  - The copter will first ascend to minimum safety altitude of 1 meters above the ground or home, depending on the presence of a distance sensor.
  - It will ascend until it is 3 meters within the [follow height](#FLW_TGT_HT) to avoid potential collisions before moving horizontally.
  - Copter will always adjust it's heading to face the target

At this point you can start moving, and the drone will be following you.

The mode has been tested on the following Android devices:
- Galaxy S10
- Nexus 7 Tablet

## Using MAVSDK to use Follow Me

> MAVSDK supports [Follow Me](https://mavsdk.mavlink.io/main/en/cpp/guide/follow_me.html). However it currently has a bug where a same message is getting sent twice occasionally, which can mess with the target position and velocity estimator.

Reviewer: Provide a bug link here

Author: I have reported just now : https://github.com/mavlink/MAVSDK/issues/1756
Hoping it will get solved before merging this doc :pray:

Reviewer: I reworded this and add the bug number. It is better to have the bug issue there and forget to remove it than merge this without adding it.

:::

The follow-me behavior can be configured using the following parameters:

Parameter | Description
--- | ---
<a id="FLW_TGT_ALT_M"></a>[FLW_TGT_ALT_M](../advanced_config/parameter_reference.md#FLW_TGT_ALT_M) | Altitude control mode. <br>- `0` = 2D Tracking <br>- `1` = 2D Tracking + Terrain Following <br>- `2` = 3D Tracking of the target's altitude
<a id="FLW_TGT_HT"></a>[FLW_TGT_HT](../advanced_config/parameter_reference.md#FLW_TGT_HT) | Vehicle follow-me height. Note that this height is fixed *relative to the home/arming position* (not the target vehicle). Default and minimum height is 8 meters (about 26 ft)
<a id="FLW_TGT_DST"></a>[FLW_TGT_DST](../advanced_config/parameter_reference.md#FLW_TGT_DST) | Vehicle/ground station separation in the *horizontal* (x,y) plane. Minimum allowed separation is 1 meter. Default distance is 8 meters (about 26 ft).
<a id="FLW_TGT_FS"></a>[FLW_TGT_FS](../advanced_config/parameter_reference.md#FLW_TGT_FS) | Flight position relative to the user when follow-me mode is active.<br>- `0` = Follow from the front right.<br>- `1` = Follow from behind or trail the user (Default).<br>- `2` = Follow from the front.<br>- `3` = Follow from the front left.

> ### Altitude Control mode explanation

Reviewer: I would move this above the parameters and have parameters as a separate heading. It is worthy of its own topic. 
```suggestion
### Altitude Control Mode

```

The altitude control mode determine whether the vehicle altitude is relative to the home position, terrain height, or the altitude reported by the follow target. 

- `2D tracking` (the default [altitude mode](#FLW_TGT_ALT_M)) makes the drone follow at a height relative to the fixed home position (takeoff altitude).
  The relative distance to the drone to the target will change as you ascend and descend (use with care in hilly terrain).

- `2D + Terrain` makes the drone follow at a fixed height relative to the terrain underneath it, using information from a distance sensor.

  - If the vehicle does not have a distance sensor following will be identical to `2D tracking`.
  - Distance sensors aren't always accurate and vehicles may be "jumpy" when flying in this mode.
  - Note that that height is relative to the ground underneath the vehicle, not the follow target.
    The drone may not follow altitude changes of the target!
  
-  `3D tracking` mode makes the drone follow at a height relative to the follow target, as supplied by its GPS sensor.
   This adapts to target altitude changes, such as when you walk up a hill.

>   :::warning

Reviewer: Moved the warning to where you're most likely to read it. The parameters section also links to this section. I also made all these modes more clear (IMO) by explicitly stating that they change the reference.

<a id="FLW_TGT_ALT_M"></a>[FLW_TGT_ALT_M](../advanced_config/parameter_reference.md#FLW_TGT_ALT_M) | Altitude control mode. <br>- `0` = 2D Tracking (Altitude Fixed) <br>- `1` = 2D Tracking + Terrain Following <br>- `2` = 3D Tracking of the target's GPS altitude **WARNING: [DO NOT USE WITH QGC for Android](#altitude-control-mode)**.
<a id="FLW_TGT_MAX_VEL"></a>[FLW_TGT_MAX_VEL](../advanced_config/parameter_reference.md#FLW_TGT_MAX_VEL) | Maximum relative velocity for orbital motion around the target, in m/s.<br>- 10 m/s has proven to be a sweet spot for aggressiveness vs smoothness.<br>- Setting it to higher value means the orbit trajectory around the target will move faster, but if the drone is physically not capable of achieving that speed, it leads to an aggressive behavior.
<a id="FLW_TGT_RS"></a>[FLW_TGT_RS](../advanced_config/parameter_reference.md#FLW_TGT_RS) | Dynamic filtering algorithm responsiveness that filters incoming target location.<br>- `0.0` = Very sensitive to movements and noisy estimates of position, velocity and acceleration.<br>- `1.0` = Very stable but not responsive filter

> ### Tips and tricks of Follow Me mode

Reviewer: The context is obvious:
```suggestion
### Tips and tricks

```

<a id="FLW_TGT_RS"></a>[FLW_TGT_RS](../advanced_config/parameter_reference.md#FLW_TGT_RS) | Dynamic filtering algorithm responsiveness that filters incoming target location.<br>- `0.0` = Very sensitive to movements and noisy estimates of position, velocity and acceleration.<br>- `1.0` = Very stable but not responsive filter

### Tips and tricks

1. Set the [follow distance](#FLW_TGT_DST) to more than 12 meters (8 meters is a "recommended minimum").

   There is an inherent position bias (3 ~ 5 meters) between the target and the drone's GPS sensor, which makes the drone follow a 'ghost target' somewhere near the actual target.
   This is more obvious when the follow distance is very small.
   We recommend that the follow distance is set to be large enough such that the GPS bias is not significant.

2. The speed at which you can change the follow angle depends on the [maximum tangential velocity](#FLW_TGT_MAX_VEL) setting.

   Experimentation shows that values between `5 m/s` are `10 m/s` are usually suitable.

> 3. Using the RC Adjustment for height, distance and angle, you can get some creative shots (using a drone with a camera)!

Reviewer: ```suggestion
3. Using the RC Adjustment for height, distance and angle, you can get some creative camera shots.

```
